### PR TITLE
Vision Subsystem Commands

### PR DESCRIPTION
**Pull request by deadlyvortex**

### CODE REVIEW

**Comment by deadlyvortex** [comment]
close #15

**Comment by deadlyvortex** [comment]
open #15

**Comment by deadlyvortex** [comment]
open #15

**Comment by deadlyvortex** [comment]
close #15

**Review by jkleiber — CHANGES_REQUESTED** [review]
General comments:
1. You need to create the vision subsystem and vision IO subsystem first. You can copy and paste this from the beta repo
2. The file you have isn't in C++ and is located in the wrong directory
3. You need to add PhotonVision to the project as a vendor dependency. You should look up how to do this and then do it

**Comment by deadlyvortex** [comment]
open #15

**Comment by danielbrownmsm** [comment]
for some reason deadlyvortex's commits were showing up as mine, probably because that was probably the computer I used last year. That should be all git config'ed and fixed now, though

File: .vscode/settings.json
Comment (jkleiber) [inline .vscode/settings.json:18]: delete this line. Error squiggles need to be turned on. You can't just ignore your errors.

File: src/main/cpp/Subsystems/Vision.cpp
Comment (jkleiber) [inline src/main/cpp/Subsystems/Vision.cpp:36]: Fix indentation please

File: src/main/cpp/VisionIO.cpp
Comment (jkleiber) [inline src/main/cpp/VisionIO.cpp:3]: This should be called using `VOKC_CALL`

File: src/main/cpp/VisionIO.cpp
Comment (jkleiber) [inline src/main/cpp/VisionIO.cpp:31]: Clean up this formatting / indentation please. It is really hard to read this code.

File: src/main/include/io/VisionIO.h
Comment (jkleiber) [inline src/main/include/io/VisionIO.h:47]: Fix the indentation please

**Review by jkleiber — COMMENTED** [review]
@deadlyvortex This looks like it probably would build correctly if the merge conflict gets resolved. Please fix the merge conflict and address the comments about formatting (you can look at the other files in the repository for examples of decent formatting), and I think it will be good to go!

I mostly left comments about fixing formatting. This is important because code MUST be readable. In software development, you spend more time reading code than writing code. Sometimes you will be looking back on your own code and wondering what it did - so it needs to be written very clearly in order to save on headaches later.

File: src/main/cpp/VisionIO.cpp
Comment (deadlyvortex) [reply src/main/cpp/VisionIO.cpp:3]: the include file needs to be changed right

Comment (jkleiber) [reply src/main/cpp/VisionIO.cpp:3]: Yeah you would need to include Utils.h

File: .vscode/settings.json
Comment (jkleiber) [inline .vscode/settings.json:17]: Fix merge conflicts

File: .vscode/settings.json
Comment (jkleiber) [inline .vscode/settings.json:95]: Fix merge conflicts / stash

File: src/main/cpp/VisionIO.cpp
Comment (jkleiber) [inline src/main/cpp/VisionIO.cpp:16]: Fix indentation (add tabs to each level of brackets). For example:
```
bool VisionIO::ProcessIO() {
(tab) if (...) {
(tab) (tab) code inside if
(tab) }
```
please do this style for the whole file.

File: src/main/include/io/VisionIO.h
Comment (jkleiber) [inline src/main/include/io/VisionIO.h:18]: name this `vision_config_t` please

**Review by jkleiber — CHANGES_REQUESTED** [review]
Some more formatting comments. I resolved the comments that I think are done as well

**Comment by jkleiber** [comment]
Another comment @deadlyvortex - you need to change the "Subsystems" folder to be named "subsystems" (lower case name). The reason the build fails is because of this

**Review by jkleiber — APPROVED** [review]
Looks good - please merge this if it passes CI

**Comment by jkleiber** [comment]
Sorry for all the spam guys. The issue is not with our code - this is due to a known issue in the WPILib DataLogger: https://github.com/wpilibsuite/allwpilib/issues/5120

master fails to build on my Linux computer as well, so for now we will be without CI - I will lift the requirement on the CI checks passing, and we'll wait to see if they can fix this issue. For now, we will work on the honor system if it builds locally on a windows laptop.

**Review by danielbrownmsm — APPROVED** [review]
Looks about right, yeah. It might be nice to rename the software interface variables to be a little more descriptive, but I think it's probably more important to stat working on actual functionality.

**Comment by jkleiber** [comment]
@deadlyvortex since you did most of the work on this, please merge it whenever you get the chance

**Comment by gaborszita** [comment]
> Sorry for all the spam guys. The issue is not with our code - this is due to a known issue in the WPILib DataLogger: [wpilibsuite/allwpilib#5120](https://github.com/wpilibsuite/allwpilib/issues/5120)
> 
> master fails to build on my Linux computer as well, so for now we will be without CI - I will lift the requirement on the CI checks passing, and we'll wait to see if they can fix this issue. For now, we will work on the honor system if it builds locally on a windows laptop.

I'm looking into this error and see if I can fix it.

**Comment by gaborszita** [comment]
> Sorry for all the spam guys. The issue is not with our code - this is due to a known issue in the WPILib DataLogger: [wpilibsuite/allwpilib#5120](https://github.com/wpilibsuite/allwpilib/issues/5120)
> 
> master fails to build on my Linux computer as well, so for now we will be without CI - I will lift the requirement on the CI checks passing, and we'll wait to see if they can fix this issue. For now, we will work on the honor system if it builds locally on a windows laptop.

I'm looking into this error and see if I can fix it. In the meantime, try disabling network tables logging by `DataLogManager::LogNetworkTables(false)` since that removed the error for me.

**Comment by gaborszita** [comment]
> Sorry for all the spam guys. The issue is not with our code - this is due to a known issue in the WPILib DataLogger: [wpilibsuite/allwpilib#5120](https://github.com/wpilibsuite/allwpilib/issues/5120)
> 
> master fails to build on my Linux computer as well, so for now we will be without CI - I will lift the requirement on the CI checks passing, and we'll wait to see if they can fix this issue. For now, we will work on the honor system if it builds locally on a windows laptop.

Also, thank you for pointing out that the error is only present on Windows. This will make my debugging process easier.

**Comment by gaborszita** [comment]
It seems I fixed the issue. I'll create a PR and see if they merge it.

**Review by danielbrownmsm — APPROVED** [review]

**Comment by gaborszita** [comment]
FYI @jkleiber upgrading to WPLib 2023.4.2 (you can upgrade simply by changing the GradleRIO plugin version in `build.gradle`) fixes the WPILib DataLogger issue.

**Comment by jkleiber** [comment]
@gaborszita apologies for the slow response, but thanks for fixing this issue! Unbreaking our CI is greatly appreciated 😄

**Review by danielbrownmsm — DISMISSED** [review]

**Review by danielbrownmsm — COMMENTED** [review]
want to get a final review from @jkleiber too

**Review by jkleiber — APPROVED** [review]
LGTM

**Pull request merged**
